### DISCUSSION TITLE
Show line number after file path to allow directly jumping to test function

### DISCUSSION THREAD

**Discussion by harupy** [pytest-dev/pytest#8085] [Ideas]
I think it'd be useful if file path was suffixed with line number (I captured the GIF attached in VSCode). This allows us to jump to test functions easily.

```python
# test_file.py

def test_func():
    pass

def TestClass:
    def test_method(self):
        pass
```

```
> pytest test_file.py --show-lineno --verbose
...

test_file.py:0::test_func PASSED              [ 50%]
test_file.py:5::TestClass::test_method PASSED [100%]
            ^^^
            insert line number here if `show_linen` is true
...
```

![jump3](https://user-images.githubusercontent.com/17039389/100521476-30bf9380-31e7-11eb-9bbf-32339a6ab427.gif)

My fork branch containing a patch that enables this feature: https://github.com/harupy/pytest/tree/show-lineno

**Comment by harupy** [comment]
<img width="572" alt="Screen Shot 2020-11-29 at 14 50 48" src="https://user-images.githubusercontent.com/17039389/100534461-5c2b9800-3252-11eb-9376-d33887c28639.png">

PyCharm also recognizes `<filepath>:<lineno>` as a link.

**Reply by RonnyPfannschmidt** [reply]
if we want to make this consistent its a "breaking" change that alters the nodeids

im all for it, but i'd like to make it part of turning nodeids into something not quite a string

**Comment by harupy** [comment]
I figured out how to achieve this without modifying the pytest codebase.

```python
class NewTerminalReporter(TerminalReporter):
    def _locationline(self, nodeid, fspath, lineno, domain):
        res = super()._locationline(nodeid, fspath, lineno, domain)
        return res.replace("::", ":{}::".format(lineno), 1)


@pytest.hookimpl(trylast=True)
def pytest_configure(config):
    old_reporter = config.pluginmanager.getplugin("terminalreporter")
    config.pluginmanager.unregister(old_reporter)
    new_reporter = NewTerminalReporter(config, sys.stdout)
    config.pluginmanager.register(new_reporter, "terminalreporter")
```

**Comment by jace** [comment]
I got it to work by dropping this into `conftest.py`:

```python
def pytest_runtest_logreport(report: pytest.TestReport):
    """Add line numbers to log report, for easier discovery in code editors."""
    # Report location of test (this is NOT the failing line)
    report.nodeid = f'{report.location[0]}:{report.location[1]}::{report.location[2]}'
```

The reported line numbers are one less than the actual start of the test definition, usually the preceding blank line, same as what @harupy's screencap shows. It's not clear why, but it's tolerable.

When a test fails, I'd like it to report the line number of the failing line so it's easier to jump to, but for that I guess I'll have to examine the traceback and confirm the filename is the same as the test file…

**Reply by jace** [reply]
Updated for showing failing line number. This appears to be private API in Pytest, so I don't know if it's reliable across failure scenarios.

```python
def pytest_runtest_logreport(report: pytest.TestReport):
    """Add line numbers to log report, for easier discovery in code editors."""
    # Report location of test (failing line number if available, else test location)
    filename, line_no, domain = report.location
    if report.longrepr is not None:
        if (
            repr_traceback := getattr(report.longrepr, 'reprtraceback', None)
        ) is not None:
            repr_file_loc = repr_traceback.reprentries[0].reprfileloc
            if repr_file_loc.path == filename:
                line_no = repr_file_loc.lineno
    report.nodeid = f'{filename}:{line_no}::{domain}'
```
